File: client/cmd/compute/compute.go
package compute

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"slices"
	"strconv"
	"strings"
	"time"

	"github.com/pkg/errors"
	"github.com/spf13/cobra"
	"go.uber.org/zap"
	"google.golang.org/grpc"

	aliases "source.quilibrium.com/quilibrium/monorepo/alias"
	"source.quilibrium.com/quilibrium/monorepo/bls48581"
	"source.quilibrium.com/quilibrium/monorepo/bulletproofs"
	"source.quilibrium.com/quilibrium/monorepo/client/utils"
	"source.quilibrium.com/quilibrium/monorepo/config"
	"source.quilibrium.com/quilibrium/monorepo/node/keys"
	"source.quilibrium.com/quilibrium/monorepo/protobufs"
	tkeys "source.quilibrium.com/quilibrium/monorepo/types/keys"
)

var nodeConfig *config.Config
var keyMgr tkeys.KeyManager
var aliasStore *aliases.Store

func getConfig() *config.Config {
	if nodeConfig != nil {
		return nodeConfig
	}
	cfg, err := utils.LoadDefaultNodeConfig()
	if err != nil {
		return nil
	}
	nodeConfig = cfg
	return cfg
}

func initKeyManager() {
	if keyMgr != nil {
		return
	}
	cfg := getConfig()
	if cfg == nil {
		return
	}
	keyMgr = keys.NewFileKeyManager(cfg, nil, nil, nil)
}

func initAliasStore() {
	if aliasStore != nil {
		return
	}
	cfg := getConfig()
	if cfg == nil {
		return
	}
	aliasStore, _ = utils.LoadAliasStore(cfg)
}

func resolveAddress(input string, expectedLen int) ([]byte, error) {
	initAliasStore()
	if aliasStore != nil {
		if addr, _, ok := aliasStore.Resolve(input); ok {
			if len(addr) != expectedLen {
				return nil, fmt.Errorf(
					"alias %q resolved to %d bytes, expected %d",
					input, len(addr), expectedLen,
				)
			}
			fmt.Printf("Resolved alias %q to %s\n", input, hex.EncodeToString(addr))
			return addr, nil
		}
	}

	h := strings.TrimPrefix(input, "0x")
	b, err := hex.DecodeString(h)
	if err != nil {
		return nil, fmt.Errorf("must be an alias or hex address: %w", err)
	}
	if len(b) != expectedLen {
		return nil, fmt.Errorf(
			"expected %d bytes (%d hex chars), got %d bytes",
			expectedLen, expectedLen*2, len(b),
		)
	}
	return b, nil
}

func getNodeClient() (protobufs.NodeServiceClient, *grpc.ClientConn, error) {
	cfg := getConfig()
	if cfg == nil {
		return nil, nil, errors.New("no config available")
	}

	conn, err := utils.GetGRPCClient(false, "", cfg)
	if err != nil {
		return nil, nil, errors.Wrap(err, "get node client")
	}

	return protobufs.NewNodeServiceClient(conn), conn, nil
}

var ComputeCmd = &cobra.Command{
	Use:   "compute",
	Short: "Compute operations",
	Long:  `Commands for executing compute operations on the Quilibrium network.`,
}

var ExecuteCmd = &cobra.Command{
	Use:   "execute <FullAddress|Alias> [Rendezvous] [PartyId] [ArgumentKey=ArgumentValue...]",
	Short: "Execute a compute operation",
	Long: `Performs an execution of a given address (or alias), as the given party identifier (1, 2, ..., n)
using the rendezvous address if applicable, and finalizes any settled data outputs to the hypergraph.`,
	Args: cobra.MinimumNArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		// Parse domain (full address or alias)
		domainBytes, err := resolveAddress(args[0], 32)
		if err != nil {
			return fmt.Errorf("domain: %w", err)
		}

		var domain [32]byte
		copy(domain[:], domainBytes)

		// Parse optional arguments
		var rendezvousHex, partyIdStr string
		var arguments []string
		for i := 1; i < len(args); i++ {
			if strings.Contains(args[i], "=") {
				arguments = append(arguments, args[i])
			} else if rendezvousHex == "" {
				rendezvousHex = args[i]
			} else if partyIdStr == "" {
				partyIdStr = args[i]
			}
		}

		// Parse or generate rendezvous
		var rendezvous [32]byte
		if rendezvousHex != "" {
			rendezvousBytes, err := hex.DecodeString(strings.TrimPrefix(rendezvousHex, "0x"))
			if err != nil {
				return fmt.Errorf("invalid rendezvous hex: %w", err)
			}
			if len(rendezvousBytes) != 32 {
				return fmt.Errorf("rendezvous must be 32 bytes, got %d bytes", len(rendezvousBytes))
			}
			copy(rendezvous[:], rendezvousBytes)
		} else {
			if _, err := rand.Read(rendezvous[:]); err != nil {
				return fmt.Errorf("generate rendezvous: %w", err)
			}
		}

		// Parse party ID
		_ = partyIdStr // party ID is used for multi-party execution context

		// Init crypto
		initKeyManager()
		if keyMgr == nil {
			return fmt.Errorf("key manager not available")
		}

		cfg := getConfig()
		if cfg == nil {
			return fmt.Errorf("no config available")
		}

		logger, _ := zap.NewProduction()
		_ = bls48581.NewKZGInclusionProver(logger) // inclusionProver for future use
		bulletproofProver := bulletproofs.NewBulletproofProver()

		// Get peer private key for proof of payment
		peerPrivKey, err := hex.DecodeString(cfg.P2P.PeerPrivKey)
		if err != nil {
			return fmt.Errorf("decode peer private key: %w", err)
		}

		payerPublicKey := peerPrivKey[57:]
		secretKey := peerPrivKey[:57]

		// Build execute operations
		operations := []*protobufs.ExecuteOperation{}
		mainOp := &protobufs.ExecuteOperation{
			Application: &protobufs.Application{
				Address:          domain[:],
				ExecutionContext: protobufs.ExecutionContext_EXECUTION_CONTEXT_HYPERGRAPH,
			},
		}

		// Add argument identifier if provided
		if len(arguments) > 0 {
			var identBytes []byte
			for _, arg := range arguments {
				identBytes = append(identBytes, []byte(arg)...)
			}
			mainOp.Identifier = identBytes
		} else {
			// Use party ID as identifier if available
			if partyIdStr != "" {
				partyID, err := strconv.ParseUint(partyIdStr, 10, 32)
				if err != nil {
					return fmt.Errorf("invalid party ID: %w", err)
				}
				mainOp.Identifier = []byte(fmt.Sprintf("party_%d", partyID))
			} else {
				mainOp.Identifier = []byte("default")
			}
		}

		operations = append(operations, mainOp)

		// Build proof of payment
		proofOfPayment := [][]byte{
			payerPublicKey,
			bulletproofProver.SimpleSign(secretKey, rendezvous[:]),
		}

		// Build CodeExecute message
		codeExecute := &protobufs.CodeExecute{
			ProofOfPayment:    proofOfPayment,
			Domain:            domain[:],
			Rendezvous:        rendezvous[:],
			ExecuteOperations: operations,
		}

		request := &protobufs.MessageRequest{
			Request: &protobufs.MessageRequest_CodeExecute{
				CodeExecute: codeExecute,
			},
		}

		// Send message
		client, conn, err := getNodeClient()
		if err != nil {
			return fmt.Errorf("connect to node: %w", err)
		}
		defer conn.Close()

		bundle := &protobufs.MessageBundle{
			Requests:  []*protobufs.MessageRequest{request},
			Timestamp: time.Now().UnixMilli(),
		}

		payload, err := bundle.ToCanonicalBytes()
		if err != nil {
			return fmt.Errorf("serialize message: %w", err)
		}

		signer, err := keyMgr.GetSigningKey("q-peer-key")
		if err != nil {
			return fmt.Errorf("get signing key: %w", err)
		}

		sig, err := signer.SignWithDomain(
			payload,
			slices.Concat([]byte("NODE_AUTHENTICATION"), domain[:]),
		)
		if err != nil {
			return fmt.Errorf("sign message: %w", err)
		}

		_, err = client.Send(
			context.Background(),
			&protobufs.SendRequest{
				Domain:         domain[:],
				Request:        bundle,
				Authentication: sig,
			},
		)
		if err != nil {
			return fmt.Errorf("send code execute: %w", err)
		}

		fmt.Printf("Code execution submitted successfully\n")
		fmt.Printf("Rendezvous: %s\n", hex.EncodeToString(rendezvous[:]))

		return nil
	},
}

func init() {
	ComputeCmd.AddCommand(ExecuteCmd)
}

File: crates/quil-client/src/commands/compute.rs
//! `qclient compute execute <FullAddress|Alias> [Rendezvous] [PartyId] [k=v...]`.
//!
//! Port of `client/cmd/compute/compute.go` `ExecuteCmd`, rebuilt for PQ:
//! the proof-of-payment is a **Falcon** signature over the rendezvous
//! (domain-separated by the compute domain) — `[payer_pubkey, sig]` — the
//! post-quantum replacement for the Go decaf `SimpleSign`
//! (`compute_intrinsic/intrinsic.rs::verify_code_execute`).

use clap::{Args, Subcommand};
use rand::RngCore;

use quil_keys::FileKeyManager;
use quil_types::crypto::Signer;
use quil_types::proto::compute::{
    Application, CodeExecute, ExecuteOperation, ExecutionContext,
};
use quil_types::proto::global::{message_request::Request, MessageRequest};

use crate::alias_store::{self, Store};
use crate::context::{Context, GlobalArgs};
use crate::rpc::ConnectOpts;

/// Build a `CodeExecute` op for `domain` with the given operations. The
/// proof-of-payment is a Falcon `q-prover-key` signature over `rendezvous`
/// (FN-DSA context = `domain`), packaged as `[payer_pubkey, sig]` — exactly
/// what the node's `verify_code_execute` checks.
pub(crate) fn build_code_execute(
    key_manager: &FileKeyManager,
    domain: &[u8],
    rendezvous: &[u8],
    execute_operations: Vec<ExecuteOperation>,
) -> anyhow::Result<CodeExecute> {
    let signer: Box<dyn Signer> = key_manager
        .get_signer_by_id("q-prover-key")
        .map_err(|e| anyhow::anyhow!("get payer key (q-prover-key): {e}"))?;
    let payer = signer.public_key().to_vec();
    let payment_sig = signer
        .sign_with_domain(rendezvous, domain)
        .map_err(|e| anyhow::anyhow!("sign proof of payment: {e}"))?;

    Ok(CodeExecute {
        proof_of_payment: vec![payer, payment_sig],
        domain: domain.to_vec(),
        rendezvous: rendezvous.to_vec(),
        execute_operations,
    })
}

#[derive(Debug, Args)]
pub struct ComputeArgs {
    #[command(subcommand)]
    pub command: ComputeCommand,
}

#[derive(Debug, Subcommand)]
pub enum ComputeCommand {
    /// Execute a compute operation on a domain.
    Execute {
        /// Domain full address or alias (32-byte app address).
        address: String,
        /// Optional 32-byte hex rendezvous (random if omitted), party id,
        /// and `key=value` arguments.
        rest: Vec<String>,
    },
}

pub async fn run(global: GlobalArgs, args: &ComputeArgs) -> anyhow::Result<()> {
    match &args.command {
        ComputeCommand::Execute { address, rest } => execute(global, address, rest).await,
    }
}

fn resolve(store: &Option<Store>, input: &str, expected: usize) -> anyhow::Result<Vec<u8>> {
    if let Some(s) = store {
        if let Some((addr, _)) = s.resolve(input) {
            if addr.len() != expected {
                anyhow::bail!("alias {input:?} resolved to {} bytes, expected {expected}", addr.len());
            }
            return Ok(addr);
        }
    }
    let b = hex::decode(input.strip_prefix("0x").unwrap_or(input))
        .map_err(|e| anyhow::anyhow!("must be an alias or hex address: {e}"))?;
    if b.len() != expected {
        anyhow::bail!("expected {expected} bytes, got {}", b.len());
    }
    Ok(b)
}

async fn execute(global: GlobalArgs, address: &str, rest: &[String]) -> anyhow::Result<()> {
    let ctx = Context::load(global)?;
    let (node_config, dir) = ctx.load_node_config("default")?;
    let alias_store = alias_store::try_load_for_config_dir(&dir);
    let key_manager = ctx.key_manager(&node_config, &dir)?;

    let domain = resolve(&alias_store, address, 32)?;

    // Parse optional [rendezvous] [partyId] [k=v...].
    let mut rendezvous_hex: Option<String> = None;
    let mut party_id: Option<String> = None;
    let mut arguments: Vec<String> = Vec::new();
    for a in rest {
        if a.contains('=') {
            arguments.push(a.clone());
        } else if rendezvous_hex.is_none() {
            rendezvous_hex = Some(a.clone());
        } else if party_id.is_none() {
            party_id = Some(a.clone());
        }
    }

    let rendezvous: Vec<u8> = match rendezvous_hex {
        Some(h) => {
            let b = hex::decode(h.strip_prefix("0x").unwrap_or(&h))
                .map_err(|e| anyhow::anyhow!("invalid rendezvous hex: {e}"))?;
            if b.len() != 32 {
                anyhow::bail!("rendezvous must be 32 bytes, got {}", b.len());
            }
            b
        }
        None => {
            let mut r = [0u8; 32];
            rand::thread_rng().fill_bytes(&mut r);
            r.to_vec()
        }
    };
    // The reserved metadata address [0xFF;32] is rejected by the node.
    if rendezvous.iter().all(|&b| b == 0xFF) {
        anyhow::bail!("rendezvous collides with the reserved metadata address");
    }

    // Identifier: concatenated k=v args, else party id, else "default".
    let identifier: Vec<u8> = if !arguments.is_empty() {
        arguments.concat().into_bytes()
    } else if let Some(p) = &party_id {
        let n: u64 = p.parse().map_err(|_| anyhow::anyhow!("invalid party ID: {p}"))?;
        format!("party_{n}").into_bytes()
    } else {
        b"default".to_vec()
    };

    let main_op = ExecuteOperation {
        application: Some(Application {
            address: domain.clone(),
            execution_context: ExecutionContext::Hypergraph as i32,
        }),
        identifier,
        dependencies: Vec::new(),
    };

    // Falcon proof-of-payment: [payer_pubkey, sign(rendezvous, ctx=domain)].
    let op = build_code_execute(&key_manager, &domain, &rendezvous, vec![main_op])?;

    let connect_opts = ConnectOpts {
        public_rpc: false,
        custom_rpc: String::new(),
        listen_grpc_multiaddr: node_config.listen_grpc_multiaddr.clone(),
    };
    let mut client = crate::rpc::connect_node_service(&connect_opts).await?;
    let request = MessageRequest {
        request: Some(Request::CodeExecute(op)),
        timestamp: 0,
    };
    crate::send::send_message_request(&mut client, &key_manager, domain, request).await?;

    println!("Compute execution submitted successfully");
    println!("Rendezvous: {}", hex::encode(&rendezvous));
    Ok(())
}
